pycloudlib/azure/cloud.py

```diff
@@ -0,0 +1,644 @@
+# This file is part of pycloudlib. See LICENSE file for license information.
+"""Azure Cloud type."""
+import base64
+
+from azure.mgmt.resource import ResourceManagementClient
+from azure.mgmt.network import NetworkManagementClient
+from azure.mgmt.compute import ComputeManagementClient
+
+import pycloudlib.azure.util as util
+
+from pycloudlib.cloud import BaseCloud
+from pycloudlib.azure.instance import AzureInstance
+from pycloudlib.key import KeyPair
+from pycloudlib.util import create_tag
+
+
+class Azure(BaseCloud):
+    """Azure Cloud Class."""
+
+    _type = 'azure'
+
+    UBUNTU_RELEASE = {
+        "trusty": "Canonical:UbuntuServer:14.04.0-LTS",
+        "xenial": "Canonical:UbuntuServer:16.04-DAILY-LTS",
+        "bionic": "Canonical:UbuntuServer:18.04-DAILY-LTS",
+        "focal": "Canonical:0001-com-ubuntu-server-focal-daily:20_04-daily-lts"
+    }
+
+    def __init__(self, tag, client_id=None, client_secret_id=None,
+                 subscription_id=None, tenant_id=None):
+        """Initialize the connection to Azure.
+
+        Azure will try to read user credentials form the /home/$USER/.azure
+        folder. However, we can overwrite those credentials with the provided
+        id parameters.
+
+        Args:
+            tag: string used to name and tag resources with
+            client_id: user's client id
+            client_secret_id: user's client secret access key
+            subscription_id: user's subscription id key
+            tenant_id: user's tenant id key
+        """
+        super().__init__(tag)
+        self._log.debug('logging into Azure')
+        self.location = "centralus"
+        self.username = "ubuntu"
+
+        self.registered_instances = {}
+        self.registered_images = {}
+
+        config_dict = {}
+
+        if client_id:
+            config_dict["client_id"] = client_id
```

Comment:
When setting up config_dict here, let's either:
 1.  actually use the expected camelCaseKeyNames (like clientId, subscriptionId)  so we don't have to remap them in azure.util.get_client
   OR 
 2. pass those client_id, subscription_id params directly into the call signature of get_client since that's the only information we are providing via config_dict. 

Ultimately what this init translates into is either you need az cli config available, or you have to provide all of client_id, subscription_id, client_secret_id, and tenant_id right? so in either get_client , or AzureCloud we should check that config condition and raise an appropriate error.

Reply:
I like the first idea better, since I think it is cleaner to check if any of the variables is None using the set logic. I will add it to the code